From: Frank Filz <ffilz@us.ibm.com>
Subject: Fix issues with POSIX->NFSv4 ACL conversion
References: bnc#535890

1. GROUP@ Allow entry [accidentally has] NFS4_ACE_IDENTIFIER_GROUP, This
appears to have been introduced by accident as part of commit bec50c4

2. The group deny entries end up denying tcy even though tcy was just
allowed by the allow entry. This appears to be due to:
	ace->access_mask = mask_from_posix(deny, flags);
instead of:
	ace->access_mask = deny_mask_from_posix(deny, flags);

Frank

Signed-off-by: Frank Filz <ffilzlnx@...>
Acked-by: J. Bruce Fields <bfields@fieldses.org>
Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

diff -X ignore -ruNp linux-2.6.27.19-5/fs/nfsd/nfs4acl.c linux-2.6.27.19-5.5408/fs/nfsd/nfs4acl.c
--- linux-2.6.27.19-5/fs/nfsd/nfs4acl.c	2008-10-09 15:13:53.000000000 -0700
+++ linux-2.6.27.19-5.5408/fs/nfsd/nfs4acl.c	2009-08-28 13:32:40.000000000 -0700
@@ -321,7 +321,7 @@ _posix_to_nfsv4_one(struct posix_acl *pa
 	deny = ~pas.group & pas.other;
 	if (deny) {
 		ace->type = NFS4_ACE_ACCESS_DENIED_ACE_TYPE;
-		ace->flag = eflag | NFS4_ACE_IDENTIFIER_GROUP;
+		ace->flag = eflag;
 		ace->access_mask = deny_mask_from_posix(deny, flags);
 		ace->whotype = NFS4_ACL_WHO_GROUP;
 		ace++;
@@ -335,7 +335,7 @@ _posix_to_nfsv4_one(struct posix_acl *pa
 		if (deny) {
 			ace->type = NFS4_ACE_ACCESS_DENIED_ACE_TYPE;
 			ace->flag = eflag | NFS4_ACE_IDENTIFIER_GROUP;
-			ace->access_mask = mask_from_posix(deny, flags);
+			ace->access_mask = deny_mask_from_posix(deny, flags);
 			ace->whotype = NFS4_ACL_WHO_NAMED;
 			ace->who = pa->e_id;
 			ace++;
